README.md
# SqliteExt

Provides a convenient way of writing functions in Ruby that can
be called from within SQLite queries through the sqlite3 gem.

Although it is already possible to write ruby code for functions
that can be called from within SQL queries using SQLite via the
sqlite3 gem, that has some limitations which this gem seeks to
address.

First, when utilizing `SQLite3::Database#create_function`, the
added function only exists for the current instance of
`SQLite3::Database`. If that instance is being accessed through
a connection pool (e.g. via ActiveRecord) then there it is hard
to ensure that the functions are created on each new instance
before the SQL that needs them to exist is executed.

Secondly, the Ruby code needed to define a trivial SQL-callable
function using `SQLite3::Database#create_function` is fairly
verbose, and it would be nice to have a simpler API.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'sqlite_ext'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install sqlite_ext

## Usage

    SqliteExt.register_function(
      'sign',
      ->(x){ x <=> 0 }
    )

    SQLite3::Database.new 'data.db' do |db|
      puts db.execute(<<-EOS).first
        SELECT
            sign(2)
          , sign(-3)
          , sign(0)
          , COALESCE(sign(NULL), 'n/a')
      EOS
    end

    # == Output ==
    # 1
    # -1
    # 0
    # n/a

    SqliteExt.register_ruby_math

    SQLite3::Database.new 'data.db' do |db|
      puts db.execute(<<-EOS).first
        SELECT
            sqrt(25)
          , cos(asin(3.0/5.0))
          , COALESCE(sqrt(NULL), 'n/a')
      EOS
    end

    # == Output ==
    # 5.0
    # 0.8
    # n/a

## Development

After checking out the repo, run `bin/setup` to install
dependencies. Then, run `rake spec` to run the tests. You can
also run `bin/console` for an interactive prompt that will allow
you to experiment.

To install this gem onto your local machine, run `bundle exec
rake install`. To release a new version, update the version
number in `version.rb`, and then run `bundle exec rake release`,
which will create a git tag for the version, push git commits and
tags, and push the `.gem` file to
[rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at
https://github.com/stevecj/sqlite_ext. This project is intended
to be a safe, welcoming space for collaboration, and
contributors are expected to adhere to the
[Contributor Covenant](http://contributor-covenant.org) code of
conduct.


## License

The gem is available as open source under the terms of the
[MIT License](http://opensource.org/licenses/MIT).



CHANGELOG.md
### 1.5.0 - 2016-05-26
* Enhancements
  * Now works with Ruby versions prior to 2.x.

### 1.4.0 - 2016-05-26
* Enhancements
  * `SqliteExt.register_ruby_math` now also registers `pi`, `e`, `mod`,
    and `power` functions.

### 1.3.0 - 2016-05-26
* Features
  * Adds `SQLite3::Database.function_created?` to check whether a
    function with a given name has been created on the target instance.

### 1.2.0 - 2016-05-25
* Enhancements
  * `SqliteExt.register_ruby_math` no longer re-registers functions when
    they should already/still be registered.
* Features
  * `SqliteExt.register_ruby_math!` has been added to unconditionally
    register or re-register Ruby math functions.

### 1.1.0 - 2016-05-25
* Enhancements
  * `SqliteExt.register_ruby_math` now also registers `floor` and `ceil`
    functions.

### 1.0.0 - 2016-06-23
* Features
  * Added `SqliteExt.register_ruby_math`. Registers most of Ruby's `Math`
    module methods as SQLite SQL functions.

### 0.3.0 - 2016-06-23
* Removals
  * Removed support for a block argument to `SqliteExt.register_function`.
    It now accepts a `Proc` passed as a standard argument instead.
* Enhancements
  * Functions registered using `SqliteExt.register_function` automatically
    propagate `NULL` when `NULL` is passed for one or more of the required
    parameters of the `Proc` that performs the function.


spec/spec_helper.rb
spec_path = File.dirname(__FILE__)
root = File.dirname(spec_path)
lib_path = File.join(root, 'lib')
tmp_db_path = File.join(root, 'tmp', 'db')

$LOAD_PATH.unshift lib_path
require 'sqlite_ext'

require 'fileutils'
FileUtils.mkdir_p tmp_db_path

TEST_DB_FILE = File.join(tmp_db_path, 'test.db')


spec/sqlite_ext_spec.rb
require 'spec_helper'

describe SqliteExt do
  attr_accessor :db

  before do
    subject.purge_function_registrations
  end

  after do
    db.close if db && (! db.closed?)
  end

  it "has a version number" do
    expect(SqliteExt::VERSION).not_to be nil
  end

  it "allows registering a function to be auto-created using SQLite3#create_function protocol" do
    subject.register_create_function "sqrt", 1 do |fn,x|
      fn.result = Math.sqrt(x)
    end

    actual = nil
    SQLite3::Database.new(TEST_DB_FILE) do |db|
      actual = db.execute("SELECT sqrt(25)")
    end

    self.db = SQLite3::Database.new(TEST_DB_FILE)
    actual = db.execute("SELECT sqrt(25)")
    expect( actual ).to eq( [[5.0]] )
  end

  it "provides a collection of registered function names" do
    subject.register_create_function 'foo', 1 do |fn,x| ; end
    subject.register_create_function 'bar', 1 do |fn,x| ; end
    expect( subject.registered_function_names ).
      to contain_exactly( 'foo', 'bar' )
  end

  describe "registration of Ruby math" do
    before do
      subject.register_ruby_math
    end

    it "registers singleton methods from Ruby's `Math` module as functions" do
      actual = nil
      SQLite3::Database.new(TEST_DB_FILE) do |db|
        actual = db.execute(<<-EOS).first
          SELECT
            cbrt(125),
            log(1000, 10)
        EOS
      end

      expect( actual[0] ).to be_within( 0.00001 ).of( 5.0 )
      expect( actual[1] ).to be_within( 0.00001 ).of( 3.0 )
    end

    it "registers `pi` and `e` functions returning math constants" do
      actual = nil
      SQLite3::Database.new(TEST_DB_FILE) do |db|
        actual = db.execute(<<-EOS).first
          SELECT
            pi(),
            e()
        EOS
      end

      expect( actual[0] ).to be_within( 0.00001 ).of( 3.14159 )
      expect( actual[1] ).to be_within( 0.00001 ).of( 2.71828 )
    end

    it "registers `floor`, `ceil`, `mod`, and `power` functions" do
      actual = nil
      SQLite3::Database.new(TEST_DB_FILE) do |db|
        actual = db.execute(<<-EOS).first
          SELECT
            floor( 1.9 ),
            floor(-1.9 ),
            ceil( 1.1 ),
            ceil(-1.1 ),
            mod( 101.5, 100 ),
            mod( -97.5, 100 ),
            power( 2, 3 ),
            power( 3, 2 )
        EOS
      end

      expect( actual ).to eq([
         1.0,
        -2.0,
         2.0,
        -1.0,
         1.5,
         2.5,
         8.0,
         9.0
      ])
    end

    it "registers `power` function" do
      actual = nil
      SQLite3::Database.new(TEST_DB_FILE) do |db|
        actual = db.execute(<<-EOS).first
          SELECT
            power( 2, 3 ),
            power( 3, 2 )
        EOS
      end

      expect( actual ).to eq([
        8.0,
        9.0
      ])
    end
  end

  describe "registration of a function for a block that returns a value" do
    before do
      subject.register_function("format", ->(template, v1, *vv){
        vv.map!{ |v| v || -1 }
        template % [v1, *vv]
      })
      self.db = SQLite3::Database.new(TEST_DB_FILE)
    end

    it "uses the specified function to process non-NULL values" do
      actual = db.execute(
        "SELECT format('result: %d %d', 8, 9)"
      )[0][0]

      expect( actual ).to eq('result: 8 9')
    end

    it "propagates NULLs for any required arguments" do
      actual = db.execute( <<-EOS )[0]
        SELECT
            format( NULL,             NULL )
          , format( 'result: %d, %d', NULL )
          , format( NULL,             5    )
      EOS

      expect( actual ).to eq( [nil, nil, nil] )
    end

    it "processes NULLs for any optional arguments" do
      actual = db.execute( <<-EOS )[0]
        SELECT
          format('result: %d %d', 8, NULL),
          format('result: %d %d %d', 8, NULL, NULL)
      EOS

      expect( actual ).to eq([
        'result: 8 -1',
        'result: 8 -1 -1'
      ])
    end
  end

end


spec/sqlite3/database_spec.rb
require 'spec_helper'

# Functionality added to SQLite3::Database when the sqlite_ext
# gem is loaded.

describe SQLite3::Database do
  subject{ described_class.new(TEST_DB_FILE) }

  after do
    subject.close unless subject.closed?
  end

  context "when a specific function HAS NOT been created" do
    it "indicates that the function HAS NOT been created" do
      expect( subject.function_created?( :foo  ) ).to eq( false )
      expect( subject.function_created?( :FOO  ) ).to eq( false )
      expect( subject.function_created?( 'foo' ) ).to eq( false )
      expect( subject.function_created?( 'FOO' ) ).to eq( false )
    end
  end

  context "when a specific function HAS been created" do
    before do
      subject.create_function('foo', 0){ |fn|  }
    end

    it "indicates that the function HAS been created" do
      expect( subject.function_created?( :foo  ) ).to eq( true )
      expect( subject.function_created?( :FOO  ) ).to eq( true )
      expect( subject.function_created?( 'foo' ) ).to eq( true )
      expect( subject.function_created?( 'FOO' ) ).to eq( true )
    end
  end

end


lib/sqlite_ext/db_method_prepends.rb
require 'set'

module SqliteExt

  module DbMethodPrepends

    # Adds functions registered with SqliteExt to each new
    # instance before it is returned from `.new` or `.open`or is
    # passed to the given block.
    def initialize(file, *other_init_args)
      if block_given?
        super file, *other_init_args do
          SqliteExt.enhance_db_session self
          yield self
        end
      else
        super
        SqliteExt.enhance_db_session self
      end
    end

    def create_function(name, arity, *other_args, &block)
      super
      created_function_names << name_key_from(name)
    end

    # Given a name, returns true if a function of that hane has
    # been created on the target instance. The name lookup is
    # case-insensitive, and either a string or a symbol may be
    # supplied.
    def function_created?(name)
      name_key = name_key_from(name)
      created_function_names.include?(name_key)
    end

    private

    def name_key_from(name)
      "#{name}".upcase
    end

    def created_function_names
      @created_function_names ||= Set.new
    end

  end

end


lib/sqlite_ext/db_tracks_created_functions.rb
require 'set'

module SqliteExt

  module DbTracksCreatedFunctions

    # Adds recording of names of created functions to
    # `create_function`.
    if RUBY_VERSION.split('.').first.to_i >= 2

      def create_function(name, arity, *other_args, &block)
        super
        created_function_names << name_key_from(name)
      end

    else

      def self.included(other)
        orig_create_function = other.instance_method(:create_function)
        other.send :define_method, :create_function, proc{ |name, arity, *other_args, &block|
          orig_create_function.bind(self).call name, arity, *other_args, &block
          created_function_names << name_key_from(name)
        }
      end

    end

    # Given a name, returns true if a function of that hane has
    # been created on the target instance. The name lookup is
    # case-insensitive, and either a string or a symbol may be
    # supplied.
    def function_created?(name)
      name_key = name_key_from(name)
      created_function_names.include?(name_key)
    end

    private

    def name_key_from(name)
      "#{name}".upcase
    end

    def created_function_names
      @created_function_names ||= Set.new
    end

  end

end


lib/sqlite_ext/db_auto_creates_registered_functions.rb
require 'set'

module SqliteExt

  module DbAutoCreatesRegisteredFunctions

    # Adds functions registered with SqliteExt to each new
    # instance before it is returned from `.new` or `.open` or is
    # passed to the given block.
    if RUBY_VERSION.split('.').first.to_i >= 2

      def initialize(file, *other_init_args)
        if block_given?
          super file, *other_init_args do
            SqliteExt.enhance_db_session self
            yield self
          end
        else
          super
          SqliteExt.enhance_db_session self
        end
      end

    else

      def self.included(other)
        orig_initialize = other.instance_method(:initialize)

        other.send :define_method, :initialize, proc{ |file, *other_init_args, &block|
          if block
            orig_initialize.bind(self).call file, *other_init_args do
              SqliteExt.enhance_db_session self
              block.call self
            end
          else
            orig_initialize.bind(self).call file, *other_init_args
            SqliteExt.enhance_db_session self
          end
        }
      end

    end

  end

end


lib/sqlite_ext.rb
gem "sqlite3"
require "sqlite3"

module SqliteExt
  require "sqlite_ext/version"
  require "sqlite_ext/db_tracks_created_functions"
  require "sqlite_ext/db_auto_creates_registered_functions"
end

class SQLite3::Database
  if RUBY_VERSION.split('.').first.to_i >= 2
    puts 'Ruby >= 2'
    prepend SqliteExt::DbTracksCreatedFunctions
    prepend SqliteExt::DbAutoCreatesRegisteredFunctions
  else
    puts 'Ruby < 2'
    include SqliteExt::DbTracksCreatedFunctions
    include SqliteExt::DbAutoCreatesRegisteredFunctions
  end
end

module SqliteExt

  class << self

    # Registers a Ruby `Proc` to be used as a function in SQL
    # code executed through subsequent new instances of
    # `SQLite3::Database`.
    #
    # When `NULL` value is passed 1 or more of the `Proc`'s
    # required parameters, it will be "propagated", meaning that
    # the specified `Proc` will not be invoked and `NULL` will be
    # returned from the call in SQL.
    #
    # When non-`NULL` values are passed to all of the `Proc`'s
    # required parameters and `NULL` is passed to any or all of
    # the `Proc`'s optional parameters, they will NOT be
    # propagated, and the `Proc` WILL be invoked. Values passed
    # as `NULL` in SQL will be forwarded to the `Proc` as Ruby
    # `nil`s.
    #
    # Whenever the `Proc` is called and returns `nil`, that will
    # result in `NULL` being returned from the function call in
    # SQL.
    #
    # Example:
    #
    # SqliteExt.register_function(
    #   'sqrt',
    #   ->(x){ Math.sqrt(x) }
    # )
    #
    # SQLite3::Database.new 'data.db' do |db|
    #   puts db.execute(
    #     "SELECT sqrt(25), COALESCE(sqrt(NULL), -1)"
    #   ).first
    # end
    #
    # # == Output ==
    # # 5.0
    # # -1
    #
    def register_function(name, prok)
      minimum_arity =
        prok.
        parameters.select{ |(kind,_)| kind == :req }.
        count

      register_create_function name, minimum_arity do |fn,*args|
        fn.result =
          if args[0...minimum_arity].any?{ |a| a.nil? }
            nil
          else
            prok.call(*args)
          end
      end
    end

    # Registers most of the public module methods of Ruby's
    # `Math` module as well as `ceil`, `floor`, `mod`, and
    # `power` based on `Float` instance methods to be used as
    # functions in SQL code executed through subsequent new
    # instances of `SQLite3::Database`.
    #
    # The `Math.frexp` method is omitted becuse it returns an
    # array, and there is no way to return an array from a SQL
    # function in SQLite.
    #
    # `NULL`s are propagated as described in the documentation
    # for `register_function`.
    #
    # Note that calling `register_ruby_math` more than once
    # without calling `purge_function_registrations` in between
    # has no effect. Ruby math functions remain registered and
    # are not re-registered in that case.
    def register_ruby_math
      return if ruby_math_is_registered
      register_ruby_math!
    end

    # Registers or re-registers Ruby math. You may want to call
    # this method if one or more of the functions defined by a
    # previous call to `register_ruby_math` or
    # `register_ruby_math!` may have been subsequently replaced.
    #
    # See `register_ruby_math`.
    def register_ruby_math!
      fn_methods = Math.public_methods - (Module.instance_methods << :frexp)
      fn_methods.each do |m|
        register_function m, Math.method(m)
      end
      register_function 'pi', ->(){ Math::PI }
      register_function 'e',  ->(){ Math::E }
      [
        [ :floor ],
        [ :ceil  ],
        [ :modulo, :mod   ],
        [ :**,     :power ]
      ].each do |(*args)|
        m, fn = args.first, args.last
        register_function fn, m.to_proc
      end
      self.ruby_math_is_registered = true
    end

    # Registers a #create_function call to be invoked on every
    # new instance of `SQLite3::Database` immidately after it is
    # instantiated and before it is returned from the call to
    # `.new` and before the invocation of a block that is passed
    # to `.new`.
    #
    # The parameters passed to `#register_create_function` are
    # exactly the same as those that would be passed to
    # `SQLite3::Database#create_function`.
    #
    # Note that this only affects instances of
    # `SQLite3::Database` that are subsequently created and has
    # no effect on previously created instances.
    #
    # Example:
    #
    # SqliteExt.register_create_function 'sqrt', 1 do |fn,x|
    #    fn.result =
    #      case x
    #        when nil then nil
    #        else Math.sqrt(x)
    #        end
    #      end
    #
    # SQLite3::Database.new 'data.db' do |db|
    #   puts db.execute("SELECT sqrt(25)")[0][0]
    # end
    # # Output: 5.0
    #
    def register_create_function(name, arity, *other_args, &block)
      name = "#{name}"
      registered_function_creations[name] = [
        [name, arity, *other_args],
        block
      ]
    end

    # Returns an array of the names of all currently registered
    # functions.
    def registered_function_names
      registered_function_creations.keys
    end

    # Removes all function registrations. Has no effect on
    # existing instances of `SQLite3::Database`.
    def purge_function_registrations
      registered_function_creations.clear
      self.ruby_math_is_registered = false
    end

    # Creates all of the registered functions on an instance of
    # `SQLite3::Database`.
    #
    # This is normally called automatically for each new
    # instance, but it can also be used to add the functions to
    # an instance that was created before the functions were
    # registered.
    def enhance_db_session(db)
      registered_function_creations.each_value do |(args,block)|
        db.create_function *args, &block
      end
    end

    private

    attr_accessor :ruby_math_is_registered

    def registered_function_creations
      @registered_function_creations ||= {}
    end

  end
end
